### PR TITLE
Feat/derived columns

### PR DESCRIPTION
**Pull request by macintoshpie**
#### Any background context you want to provide?
Currently, users can import data into SEED into the predefined core columns (e.g. Site EUI) or into extra data (e.g. My Custom Column) for an inventory.

#### What's this PR do?
Adds initial models, DerivedColumns, which will allow users to define expressions with parameters derived from inventory values.

For example, `$a / $b * 100`, where `$a` might reference  the column "Conditioned floor area" and `$b` might reference "Total floor area".

This implementation allows numerous operators/functions (see EXPRESSION_GRAMMAR in derived_columns.py). Currently expressions can reference any `Column`. If an inventory has missing or invalid data used in the expression, it should return `None`.

This PR only adds the basic models for storing these expressions as well as code for evaluating/validating them. I.e. there is no API for creating/reading/updating/deleting these models.

#### How should this be manually tested?
Nothing for now, see unit tests.

#### What are the relevant tickets?
#1642

### CODE REVIEW

File: requirements/base.txt [hunk 1]
```diff
@@ -59,6 +59,7 @@ requests==2.22.0
 lxml==4.6.3
 probablepeople==0.5.4
 xmlschema==1.1.1
+lark==0.11.3
```
Comment (macintoshpie) [inline requirements/base.txt:62]: package for parsing/evaluating expression strings

File: seed/tests/test_derived_columns.py [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+            atomic_no_params_st | st.sampled_from(template_params),
+            recursive_st_func
+        )
+    )
+
+    return expression, params
+
+
+class TestExpressionEvaluator(TestCase):
+    @no_deadline
```
Comment (macintoshpie) [inline seed/tests/test_derived_columns.py:166]: @adrian-lara 
```suggestion
class TestExpressionEvaluator(TestCase):
    """NOTE:
    This collection of tests uses the hypothesis testing library
    See: https://hypothesis.readthedocs.io/en/latest/index.html

    Instead of making assertions about specific data, we tell hypothesis what the data _basically_
    looks like, and it runs our tests many times over different types of data it generates.
    This allows us to test cases we would have never though of and find edge cases.

    We create data specifications (ie what the data should look like) with "strategies".
    We have written custom strategies for generating different types of expressions above.
    They all end with the `_st` suffix.
    We then tell hypothesis which strategy to use for a test (ie what the data input is) by using the
    `@given` decorator.
    """

    @no_deadline
```

File: seed/tests/test_derived_columns.py [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+        derived_column = models['derived_column']
+        property_state = models['property_state']
+
+        # -- Act
+        result = derived_column.evaluate(property_state)
+
+        # -- Assert
+        self.assertIsNone(result)
+
+    def test_derived_column_evaluate_falls_back_to_passed_params_if_not_in_inventory(self):
```
Comment (adrian-lara) [inline seed/tests/test_derived_columns.py:542]: What's the use case for passing in "fallback" params? It seems related to not having an inventoryState object to base values off, but it isn't immediately clear to me when that might happen. 

I have a feeling I'm missing/forgetting something notable 😅

File: seed/migrations/0138_auto_20210512_0024.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                ('organization', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='orgs.Organization')),
+            ],
+        ),
+        migrations.CreateModel(
+            name='DerivedColumnParameter',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('parameter_name', models.CharField(max_length=50)),
+                ('derived_column', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='seed.DerivedColumn')),
+                ('source_column', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='seed.Column')),
```
Comment (adrian-lara) [inline seed/migrations/0138_auto_20210512_0024.py:31]: Not sure if this is the relevant line, but a question comes to mind here: What happens if a column is deleted that was associated to a DerivedColumn via a reference from this DerivedColumnParameter?

**Review by adrian-lara — APPROVED** [review]
A few open questions that may or may not lead to code changes. In general, the overall patterns make sense to me, and I don't see anything "off" there! 👍🏽  I'll go ahead and approve in case no code changes end up being needed. 

I'm less concerned with missing an edge case, etc given we'll later implement and use this in the frontend and test it more thoroughly during that point.

*Note/reminder that if we notice any performance hits during testing, we may consider limiting the number of tests being autogenerated for the different cases.

File: seed/models/derived_columns.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+    # All parameters used in the expression must be linked to a Column through this
+    # model's `source_columns` field
+    expression = models.CharField(max_length=526, blank=False)
+    source_columns = models.ManyToManyField(Column, through='DerivedColumnParameter')
+
+    class Meta:
+        constraints = [
+            # name must be unique
+            models.UniqueConstraint(
+                fields=['name'], name='unique_name'
```
[the remark this answers is not in the document]
Comment (macintoshpie) [reply seed/models/derived_columns.py:207]: Oops, good catch!

File: seed/tests/test_derived_columns.py (hunk 3 above)
Comment (macintoshpie) [reply seed/tests/test_derived_columns.py:542]: Yeah I think I just added that without considering it too much. I think "falling back" onto a params dict is unlikely. But passing only a dict is somewhat likely. I just didn't want to make the arguments mutually exclusive, and had to decide a precedence.

File: seed/migrations/0138_auto_20210512_0024.py (hunk 4 above)
Comment (macintoshpie) [reply seed/migrations/0138_auto_20210512_0024.py:31]: Great point, I'm going to use `PROTECT` on this for now until we decide how we want to handle this.
With `CASCADE` the `evaluate` method is just going to end up returning `None` every  time b/c it's missing that parameter for all inventory.

Comment (macintoshpie) [reply seed/migrations/0138_auto_20210512_0024.py]: See
- https://github.com/SEED-platform/seed/pull/2682/files#diff-fc0d01e7c8f83cc5d37ed1e01da02c0b9c6998a458e79e9f08eccf4f32626899R31
- https://github.com/SEED-platform/seed/pull/2682/files#diff-bfbe128f201eda3032ef868db0ae6d82527e6352043c9bcbf68e97ed3b81bc16R315

File: seed/models/derived_columns.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (macintoshpie) [reply seed/models/derived_columns.py]: See
- https://github.com/SEED-platform/seed/pull/2682/files#diff-bfbe128f201eda3032ef868db0ae6d82527e6352043c9bcbf68e97ed3b81bc16R206

**Review by adrian-lara — APPROVED** [review]
Thanks for addressing the points I made on my last review! This looks good to go! 👍🏽
